dataikuapi/dss/admin.py

```diff
@@ -1143,22 +1143,103 @@ def start(self):
             raise Exception('Cluster operation failed : %s' % (json.dumps(resp.get('messages', {}).get('messages', {}))))
         return resp
 
-    def stop(self, terminate=True):
+    def stop(self, terminate=True, force_stop=False):
         """
         Stops or detaches the cluster
 
         This operation is only valid for a managed cluster.
-        :param boolean terminate: whether to delete the cluster after stopping it
+
+        :param terminate: whether to delete the cluster after stopping it
+        :type terminate: bool
+        :param force_stop: whether to try to force stop the cluster,
+            useful if DSS expects the cluster to already be stopped
+        :type force_stop: bool
         """
         resp = self.client._perform_json(
             "POST", "/admin/clusters/%s/actions/stop" % (self.cluster_id),
-            params = {'terminate':terminate})
+            params={'terminate': terminate, 'forceStop': force_stop})
         if resp is None:
             raise Exception('Env update returned no data')
         if resp.get('messages', {}).get('error', False):
             raise Exception('Cluster operation failed : %s' % (json.dumps(resp.get('messages', {}).get('messages', {}))))
         return resp
 
+    def kubectl_command(self, args):
+        """
+        Runs an arbitrary kubectl command on the cluster.
+
+        This operation is only valid for a Kubernetes cluster.
+
+        Note: this call requires an API key with DSS instance admin rights
+
+        :param args: the arguments to pass to kubectl (without the "kubectl")
+        :type args: str
+        :return: a dict containing the return value, output, and possible error output of the command
+        :rtype: dict
+        """
+        return self.client._perform_json(
+            "POST", "/admin/clusters/%s/k8s/kubectl" % self.cluster_id,
+            body={'args': args})
+
+    def delete_finished_jobs(self, delete_failed=False, namespace="", label_filter="", dry_run=False):
+        """
+        Runs a kubectl command to delete finished jobs.
+
+        This operation is only valid for a Kubernetes cluster.
+
+        :param delete_failed: if True, delete both completed and failed jobs, otherwise only delete completed jobs
+        :type delete_failed: bool
+        :param namespace: the namespace in which to delete the jobs
+        :type namespace: str
+        :param label_filter: delete only jobs matching a label filter
+        :type label_filter: str
+        :param dry_run: if True, execute the command as a "dry run"
+        :type dry_run: bool
+        :return: a dict containing the return value, output, and possible error output of the command
```

Comment:
Shouldn't we have a more structured output here ? For this, the fact that it relies on kubectl is more of an implementation detail. Ideally, we'd return something like
```
{
   success: True/False,
   deleted: [ { "namespace": "...", "id" : ...} ],
   debug: {
       stdout:
       stderr
    }
}
```

Reply:
We can add some parsing on the backend side to make the output nicer. I guess we could even use `-o name` on the delete command, but then the debug.stdout wouldn't be too interesting